Refactor Public Inputs and add them to the transcript

davidnevadoc opened the pull request:
In this PR we have added the Public Inputs to the transcript as the first step of the `prove` and `verify` algorithms. They have been added as the evaluations of the PI polynomial over the used domain (`n` field elements).

Close #133

davidnevadoc committed:
Modify `construct_dense_pi_vec`

Update `construct_dense_pi_vec` so it returns the evaluation
representation of the PI polynomial over the used domain. This means
that the result vector will always have a power of two length.

davidnevadoc committed:
Modify `construct_dense_pi_vec`

Update `construct_dense_pi_vec` so it returns the evaluation
representation of the PI polynomial over the used domain. This means
that the result vector will always have a power of two length.

davidnevadoc committed:
Add PI to the transcript

Co-authored-by: Joshua Fitzgerald <joshuabfitzgerald@gmail.com>

davidnevadoc committed:
Add PI to the transcript

Co-authored-by: Joshua Fitzgerald <joshuabfitzgerald@gmail.com>

CPerezz commented on plonk-core/src/proof_system/proof.rs:
```diff
@@ -117,6 +117,10 @@ where
 
+        for val in pub_inputs {
+            transcript.append(b"pi", val)
+        }
```
Are the `pi` recieved sparse?? 
If so, we should see if we can avoid adding the zeros and filter them out.

davidnevadoc replied:
They usually are, so now only non-zero values will be appended to the transcript.

CPerezz commented on plonk-core/src/proof_system/prover.rs:
```diff
@@ -177,6 +177,12 @@ where
+        let pi_dense = self.cs.construct_dense_pi_vec();
+        for val in &pi_dense {
+            transcript.append(b"pi", val)
+        }
```
I don't think we should append the dense repr, we should do the sparse (in case it doesn't affect the security) as the zeros will consume an insane amount of time.

davidnevadoc replied:
We could do something like this too
``` rust
for (key, val) in self.cs.public_inputs_sparse_store.iter() {
    let static_label = Box::leak(format!("pi{}", key).into_boxed_str());
    transcript.append(static_label.as_bytes(), val)
}
```
If we ensure that `public_inputs_sparse_store` has no zero values I think there would be no malleability issues.

CPerezz replied:
Well, while this works, can't we use `String` or similar?? Box leaking works but it's something better to avoid if possible.

lopeetall replied:
Can we serialize the sparse representation and append to the transcript as bytes?

lopeetall replied:
I'm AFK or else I would just check this now...instead I'll leave a comment so I don't forget to mention it: has anyone checked PlookUp branch to see if the lookup table vectors get included in the transcript?

davidnevadoc replied:
I agree with these suggestions but since `merlin` has not merged a PR in more than a year maybe we should create our own fork.

bhgomes replied:
Something to be wary of is iteration order. If you aren't using an ordered hash map then there are no guarantees on how the pi are returned from the map. You should use either an ordered map or perform a sort by index (public_input_indices) before inserting into the transcript.

davidnevadoc replied:
Finally went for a `BTreeMap` struct inside `PI` to ensure the values are returned  in order when appended to the transcript.

hdevalence replied:
> I think we should make a PR to merlin so that the label doesn't need 'static lifetime (I'm sure there was some good original reason for this, but it shouldn't be mandatory, because if the data is consumed by the hasher, it doesn't need to live that long)

Ah, no, actually, the `'static` restriction is there [on purpose](https://merlin.cool/use/protocol.html#defining-a-transcript-protocol) to prevent exactly this type of (mis)use! 

The labels are supposed to be fixed, rather than input-dependent, because the labels are used for message framing for the data that's fed into the transcript.  Merlin uses a tag-length-value format, so the transcript is guaranteed to be parseable (and hence produce a non-malleable transcript hash) if the labels are all fixed and none of them are prefixes of each other.

In this case, the correct way to feed variable data into the transcript is to use two messages, one for the application-level label, and one for the application-level data, each with their own `'static` label at the transcript level.

> I agree with these suggestions but since merlin has not merged a PR in more than a year maybe we should create our own fork.

The upstream moved to https://github.com/zkcrypto/merlin/ ; I'm happy to merge PRs there, but not this one, since the design is trying to steer you away from doing this :)

bhgomes replied:
Because the actual circuit structure (and the entire proof system actually) can be configured at runtime, the notion of "compile-time" labels is just not correct; it is defined at the wrong abstraction level. In the case of this repo, the keys are fixed at _circuit compile time_ but not at _Rust compile time_. The only way a safety feature would be useful here is if you could constrain the labels to be fixed during circuit compilation (which has nothing to do with Rust compile-time data).

hdevalence replied:
> For example, the Box::leak code above causes this guarantee to break, without any indication at all that something went wrong (other than leaking memory - but the memory guarantee is now tied to the correctness). 

Uh, sure, I guess, but the use of `Box::leak` *in a loop over user-supplied data* to intentionally leak memory in the middle of a crypto implementation seems like a pretty clear indicator that something is wrong with the implementation, no?

> Because the actual circuit structure (and the entire proof system actually) can be configured at runtime, the notion of "compile-time" labels is just not correct; it is defined at the wrong abstraction level.

No, it isn't. You can of course write an application that doesn't include all of its data in its transcript, or does so incorrectly (which is how we ended up here). There's no way for the transcript construction to solve that for you. What it can, and does, solve for you, is ensure that you only have to worry about the application-level data, and not the low-level details of ensuring that that data is unambiguously encoded in the transcript hash.

bhgomes replied:
> Uh, sure, I guess, but the use of Box::leak in a loop over user-supplied data to intentionally leak memory in the middle of a crypto implementation seems like a pretty clear indicator that something is wrong with the implementation, no?

The keys are not _user-supplied data_ in the sense of _prover-supplied_ or _verifier-supplied_, but they are _compiler-supplied_ where _compiler_ is the user defining the circuit who is allowed to do anything. I agree that `Box::leak` is bad though, we need to get rid of that regardless of the `'static` issue.

> No, it isn't. You can of course write an application that doesn't include all of its data in its transcript or does so incorrectly (which is how we ended up here). There's no way for the transcript construction to solve that for you. What it can, and does, solve for you, is ensure that you only have to worry about the application-level data, and not the low-level details of ensuring that that data is unambiguously encoded in the transcript hash.

It's the fact that the labels are `'static` that's the problem. The labels used when specifying the transcript protocol do not need to be known at _Rust compile time_ to be correct. Handling the "low-level details" in this case needs to be done by this repo; the unambiguous encoding _is defined at Rust runtime_.

And there are actually ways to have the transcript help the _circuit compile time_ enforcement of the transcript (independent of _Rust compile time_), they are just not a feature of the `merlin` project. For the long-term usefulness of this repo, we'll need to extend the functionality and create a more general tool for transcripts.

davidnevadoc committed:
Merge branch 'add-pi-transcript' of github.com:rust-zkp/ark-plonk into add-pi-transcript

davidnevadoc committed:
New public inputs implementation

davidnevadoc committed:
Change `PI` implementation

The new implmenentation uses BTreeMap to ensure no duplicates are
introduced and to easily get an ordered representation. This is
particualrly helpuful to get a unique representation to add to the
transcript.

davidnevadoc committed:
New error type `InvalidPublicInputValue`

davidnevadoc committed:
Remove leftover

davidnevadoc committed:
Add non-zero public inputs to the transcript

davidnevadoc committed:
Adapt to new `PI` struct

davidnevadoc committed:
Fix intradoc links & lint checks

davidnevadoc committed:
Add serialized `PI` to the transcript.

Public Inputs are serialized and added to the transcript under the label
"pi". The struct PI is responsible of guaranteeing a unique
representation of each possible public input vector.

davidnevadoc committed:
Remove `Deref` from `PI`

davidnevadoc committed:
Replace `PrimeField` with `FftField`.

davidnevadoc committed:
Add basic tests

bhgomes commented on plonk-core/src/circuit.rs:
```diff
@@ -480,7 +403,10 @@ mod test {
+        let (pk, vk) = circuit.compile::<PC>(&pp)?;
+
+        //Add PI to the verifier data
+        // verifier_data.pi.insert()
```
stray comment?

davidnevadoc replied:
yep 😅

bhgomes commented on plonk-core/src/constraint_system/composer.rs:
```diff
@@ -210,7 +195,7 @@ where
             q_variable_group_add: Vec::with_capacity(expected_size),
             q_lookup: Vec::with_capacity(expected_size),
-            public_inputs_sparse_store: BTreeMap::new(),
+            public_inputs: PI::new(expected_size.next_power_of_two()),
```
Does this count the padding with the extra space needed for lookups?

davidnevadoc replied:
This just depends on the `expected_size` provided at the initialization of the `StandardComposer`. If the user building the circuit knows that the lookup tables size will exceed the number of constraints he should use that value as `expected_size`. The value of the PI size will be updated with the correct size of the circuit once the circuit is finished (typically at the end of the `gadget` call).

bhgomes commented on plonk-core/src/constraint_system/helper.rs:
```diff
@@ -78,12 +79,17 @@ where
     //
     // Create a Verifier object
     let mut verifier = Verifier::new(b"demo");
+    // panic!("{:?}", verifier.cs.public_inputs);
```
Stray comment?

bhgomes commented on plonk-core/src/proof_system/pi.rs:
```diff
@@ -0,0 +1,193 @@
+#[derive(
+    CanonicalDeserialize, CanonicalSerialize, Clone, PartialEq, Debug, Hash,
+)]
+pub struct PI<F>
```
I would call this `PublicInput` or `PublicInputs`

bhgomes commented on plonk-core/src/proof_system/pi.rs:
```diff
@@ -0,0 +1,193 @@
+///  Public Inputs
+#[derive(
+    CanonicalDeserialize, CanonicalSerialize, Clone, PartialEq, Debug, Hash,
+)]
```
Safer to use `derivative::Derivative` for `Clone`, `Debug`, `Hash`, `PartialEq`

bhgomes commented on plonk-core/src/proof_system/pi.rs:
```diff
@@ -0,0 +1,193 @@
+    }
+
+    /// Updates the size of the circuit.
+    pub fn update_size(&mut self, n: usize) {
+        assert!(n.is_power_of_two());
+        self.n = n;
+    }
```
In which case do you need this? This seems unsafe.

davidnevadoc replied:
This is related with the previous comment. It is used when the final size of the circuit is greater than the initially provided `expected_value`.  Once `gadget` has been called we know the real size of the circuit and we use this to ensure the size is updated if necessary.

bhgomes commented on plonk-core/src/proof_system/pi.rs:
```diff
@@ -0,0 +1,193 @@
+    /// implicit value of empty positions.
+    /// The function will panic if an insertion is atempted in an already
+    /// occupied position.
+    pub fn insert(&mut self, pos: usize, val: F) {
+        if self.values.contains_key(&pos) {
+            panic!("Insertion in public inputs conflicts with previous value at position {}", pos);
+        }
+        if val != F::zero() {
+            self.values.insert(pos, val);
+        }
+    }
```
This does not prevent someone from doing the following:

```rust
self.insert(4, 0);
self.insert(4, 1);
```

or in the other order. Do we even want to allow this?

davidnevadoc replied:
You're right. I thought of this case but I didn't find a workaround for it as we don't keep track of `0`s. When we switch to the approach where we explicitly specify which positions are allowed to have a public input we can keep track of all insertions and prevent scenarios like this one.
The reverse would panic since the position check is done before the zero check.

bhgomes commented on plonk-core/src/proof_system/pi.rs:
```diff
@@ -0,0 +1,193 @@
+                .ok_or(Error::InvalidPublicInputValue)?;
+
+            for (pos2, elem) in item_repr.iter().enumerate() {
+                self.insert(init_pos + pos1 + pos2, *elem);
```
I think you need to accumulate as you go, this won't insert in the right place. Take for example the following structure

```
[v0, [v1, v2], v3, v4]
```

In this case the loops would put them at the following indices:

```
v0 -> 0 + 0 = 0  (element 0 of group 0)
v1 -> 1 + 0 = 1  (element 0 of group 1)
v2 -> 1 + 1 = 2  (element 1 of group 1)
v3 -> 2 + 0 = 2  (element 0 of group 2)
v4 -> 3 + 0 = 3  (element 0 of group 3)
```

In this case, we have a conflict at index `3`. Instead, you can just loop over the arrays and count as you go:

```rust
let mut count = 0;
for item in iter {
    for elem in item.to_field_elements().ok_or(Error::InvalidPublicInputValue)? {
        self.insert(init_pos + count, elem);
        count += 1;
    }
}
Ok(count)
```

davidnevadoc replied:
Good point :+1:

bhgomes commented on plonk-core/src/proof_system/proof.rs:
```diff
@@ -33,13 +33,14 @@ use ark_serialize::{
-/// formally verify that the secret witnesses used to generate the [`Proof`]
-/// satisfy a circuit that both [`Prover`](super::Prover) and
+/// It's main goal is to allow the [`Verifier`](super::Verifier) to formally
+/// verify that the secret witnesses used to generate the [`Proof`] satisfy a
```
not sure if "formally verify" is the right language here

davidnevadoc replied:
I agree, in fact I would remove all the paragraph. WDYT?

bhgomes replied:
yeah let's get rid of it

bhgomes reviewed (changes requested):
Just a few points, and a possible bug, otherwise LGTM.

davidnevadoc committed:
Stray comments

davidnevadoc committed:
Use `derivative::Derivative` instead of `derive`

davidnevadoc committed:
Drop unnecessary comment

davidnevadoc committed:
Rename `PI` to `PublicInputs`

davidnevadoc committed:
Fix `extend`

davidnevadoc committed:
Fix intradoc links

CPerezz reviewed (approved):
LGTM. 

I like that refactor of the Public Inputs TBH. Although I think we can do more in the future with that and improve them much more.

As for the concerns of @bhgomes , they're legit. But at the same time, panicking due to a PI re-assignation is not something crazy. As it means you're not building a correct circuit. Which is not worth then compiling or making proofs for.

Would like the approval of @bhgomes and maybe one more before we merge TBH as this topic lead to a broad discussion and lots of members were involved.

LukePearson1 reviewed (approved):
LGTM. Feel free to merge
